protocols/kad/src/handler.rs

```diff
@@ -601,11 +617,9 @@ impl ConnectionHandler for KademliaHandler {
     type InboundOpenInfo = ();
 
     fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
-        if self.config.allow_listening {
-            SubstreamProtocol::new(self.config.protocol_config.clone(), ())
-                .map_upgrade(Either::Left)
-        } else {
-            SubstreamProtocol::new(Either::Right(upgrade::DeniedUpgrade), ())
+        match self.mode {
```

Comment:
Can you please add some comments on what happens here?

That sounds like a rather impactful thing. Now Peer can allow and disallow accepting connections _on its own_, and I as a developer have no control over that.

Reply:
_I'll try to outline my understanding of the effects of this change. I'm happy to be wrong here, so forgive me if I am :)_

So, previously if I had `allow_listening = true`, my peer would accept incoming connections even before `Identify` happened.

Now Kademlia handshake behaviour depends on whether `Identify` already happened or not. So instead of 1 handshake on Kademlia protocol and 1 handshake on Identify, we now have (depending on handshakes order) several scenarios:

PeerA connects to PeerB. PeerA is in Server mode, PeerB has not yet learned her external addresses, so it is in Client mode.

### Scenario 1: failed connection
1. PeerA tries to connect to PeerB
2. PeerA attempts Kademlia handhsake with PeerB, PeerB rejects because Mode::Client => upgrade::DeniedUpgrade
3. PeerA performs a succesful handshake via Identify with PeerB
4. PeerB learns its external address, switches to Server mode
5. PeerB sends `ReconfigureMode` to _all its kademlia connections_, but PeerA is not among them
6. Kademlia handhsake between A and B has failed
7. Connection failed

If PeerA tries again, connection will succeed, even though _nothing has changed_ in the setup. Technically, nothing prevents PeerB of accepting PeerA's connection the first time. But now PeerA has to retry if she's sure she brought external_addresses to PeerB.

### Scenario 2: succesful connection
1. PeerA tries to connect to PeerB 
2. PeerA performs a succesful handshake via Identify with PeerB
3. PeerB switches to Server mode
4. PeerA attempts Kademlia handshake with PeerB
5. PeerB accepts
6. Connection succesful

If I understand correctly, there's no strict order in which handshakes happen, right? So it is basically a race condition.

Reply:
> PeerA attempts Kademlia handhsake with PeerB, PeerB rejects because Mode::Client => upgrade::DeniedUpgrade

Can you elaborate on how that handshake is triggered? At least `rust-libp2p` does not attempt to speak the kademlia protocol to a peer that did not advertise it first. This is the whole point of this PR: By filtering peers based on whether the advertise kademlia, we get healthier routing tables and as a result, kademlia queries finish a lot quicker. See https://github.com/libp2p/rust-libp2p/pull/3877#issuecomment-1574940336.